### PR TITLE
Refactor dispatch async handling

### LINKED ISSUE #2448: Refactor op call signatures for core and cli
Current op dispatch function signature looks like this:
```
pub type Buf = Box<[u8]>;

pub type OpAsyncFuture = Box<dyn Future<Item = Buf, Error = ()> + Send>;

pub enum Op {
  Sync(Buf),
  Async(OpAsyncFuture),
}

type DispatchFn = Fn(&[u8], Option<PinnedBuf>) -> Op;
```
I've already made a few changes towards this in #2385. Ideally the new signature would look something like this:
```
pub type Buf = Box<[u8]>;

pub type OpAsyncFuture<E> = Box<dyn Future<Item = Buf, Error = E> + Send>;

pub enum Op<E> {
  Sync(Buf),
  Async(OpAsyncFuture<E>),
}

type DispatchFn<E> = Fn(is_sync: bool, &[u8], Option<PinnedBuf>) -> Result<Op<E>, E>;

type CoreDispatchFn = DispatchFn<()>;
```
We should be able to use a variant of this for `cli` as well:
```
type CliDispatchFn = DispatchFn<DenoError>;
```
This should allow use of the try(`?`) operator, and remove the need for this wait:
https://github.com/denoland/deno/blob/60d452264198adb3da4820236cf8ea35d33486cd/cli/ops.rs#L173

cc @ry

### PR DESCRIPTION
**Pull request by afinch7**
closes #2448
This patch refactors the dispatch system to allow for better ergonomics and more simplicity

* `promise_id` is no longer exposed beyond `core::Isolate::pre_dispatch` on the rust side
* cli dispatches now return a `Result<Op<DenoError>, DenoError>` value allowing for use of the try operator and returning non-futures in the form of `Op::Sync`
* cli dispatches are now passed a new parameter `is_sync`, so they know what kind of return is expected `Op::Sync` for `is_sync = true` and `Op::Sync | Op::Async` for `is_sync = false`
* `deno::Isolate::pre_dispatch` will return a `promise_id` to js when a `Op::Async` is received

### CODE REVIEW

**Comment by ry** [comment]
Currently: An Op is either Sync or Async. If it is sync, it contains just a Buffer. If it's async it is a future which returns a buffer. Neither the sync value nor the async value have errors associated with them. Error values are encoded inside of the buffer itself. The only externally visible metadata associated with an Op is whether it is Sync or Async ([since it's defined as an enum](https://github.com/denoland/deno/blob/a71305b4febc3d8db95d3d144ae3a64c023718f0/core/isolate.rs#L28-L35)).

This patch moves the error to not be encoded in the buffer, but instead be part of the Op's metadata (or type information).

There is another piece of metadata which is currently encoded in an op's buffer, but which might be useful to extract into the type - the op id. The op id (or promise id) is [used to resolve promises on the JS side](https://github.com/denoland/deno/blob/a71305b4febc3d8db95d3d144ae3a64c023718f0/js/dispatch.ts#L37-L38).

Note that the Future trait in Rust is moving into std, and will not include an associated Error type. That is, once `std::future::Future` becomes stable, there will only be the associated `Future::Output` type.

I think it is desirable to continue to serialize errors into the Buf, and handle them so transparently in core.

I'm happy to add the `is_sync: bool` parameter to the dispatch function, and it's fine if Ops are made generic on the Error for use elsewhere, but the dispatch function should still return `Op<()>`.

File: core/isolate.rs [hunk 1]
```diff
@@ -50,11 +56,17 @@ pub enum StartupData<'a> {
   None,
 }
 
-type DispatchFn = Fn(&[u8], Option<PinnedBuf>) -> Op;
```
Comment (ry) [inline core/isolate.rs]: This can be changed to:
```rust
type DispatchFn = Fn(bool, &[u8], Option<PinnedBuf>) -> Op<()>;
```

But no errors handling.

File: core/isolate.rs [hunk 2]
```diff
@@ -208,18 +228,31 @@ impl Isolate {
...
+          isolate.pending_ops.push(fut);
+          isolate.have_unpolled_ops = true;
+        }
+        // Panic on unexpected synchronicity. Maybe it would be better to have two
+        // dispatch functions(one for sync one for async).
+        (true, Op::Async(_)) => {
+          panic!("Dispatch returned Op::Async for sync call!")
+        }
+        (false, Op::Sync(_)) => {
+          panic!("Dispatch returned Op::Sync for async call!")
```
Comment (afinch7) [inline core/isolate.rs]: The sync value here could easily be converted to a future, but should we? I already did this in `cli::ops::dispatch_all_legacy`, but I'm not sure If this should apply to core or not.

File: cli/ops.rs [hunk 3]
```diff
@@ -755,7 +776,13 @@ fn op_fetch(
           },
         ))
       });
-  Box::new(future)
+  if is_sync {
+    // TODO(afinch7) avoid block_on here?
+    let result_buf = tokio_util::block_on(future)?;
+    Ok(Op::Sync(result_buf))
```
Comment (afinch7) [inline cli/ops.rs]: Is there a more ideal way to do this? wait?

**Comment by afinch7** [comment]
Should we refactor op/promise id to be internal to core? I don't really see any good reason that needs to be exposed beyond core.

File: core/isolate.rs (hunk 1 above)
Comment (afinch7) [reply core/isolate.rs]: My intent wasn't to handle them in core, but rather just to have a common function signature. That didn't work out since we need state as a context for dispatch in cli, so I will just revert this for now.

**Comment by ry** [comment]
> Should we refactor op/promise id to be internal to core? I don't really see any good reason that needs to be exposed beyond core.

It's not used in core - but it is used in the JS side of the dispatch... So I could see benefits of having that be a core metadata - it would allow us to use a single promise table for all ops (including plugin ops). That said, it seems superfluous to expose at the moment. I brought it up just so you're aware of what I'm thinking about.

File: cli/ops.rs (hunk 3 above)
Comment (ry) [reply cli/ops.rs]: The ideal way would be if the OpCreators themselves returned Op objects. So if they returned a Op::Sync they would be forced to supply the output buf immediately. I have a half-baked patch that edits ops.rs so everyone is returning an Op.

Comment (afinch7) [reply cli/ops.rs]: Let me rephrase that a little. When we have to use some form of wait I.E. `op_fetch_module_meta_data`, is it better to use `block_on` or just `wait()`?

Comment (ry) [reply cli/ops.rs]: This is `op_fetch` ... I think it should error out if a sync op is attempted.

Ideally `op_fetch_module_meta_data` wouldn't be sync either - but it was the only way I could get the TS compiler to work.

**Comment by afinch7** [comment]
@ry I'm mostly happy with this is one now PTAL.

File: core/libdeno/binding.cc [hunk 4]
```diff
@@ -223,23 +223,29 @@ void Send(const v8::FunctionCallbackInfo<v8::Value>& args) {
 
   v8::HandleScope handle_scope(isolate);
 
+  int cmd_id = 0;
+  if (args[0]->IsNumber()) {
+    auto cmd_id_v8_handle = v8::Local<v8::Number>::Cast(args[0]);
+    cmd_id = cmd_id_v8_handle->Value();
```
Comment (ry) [inline core/libdeno/binding.cc]: This may have a perf implication. Did you see any degradation in the bench?

Comment (afinch7) [reply core/libdeno/binding.cc]: A little <= 5%

Comment (afinch7) [reply core/libdeno/binding.cc]: Would encoding this into control be more efficient?

Comment (ry) [reply core/libdeno/binding.cc]: Yea that's what currently happens.

One of the main reasons for this "op" design was the observation we had from Node, that once you're out of JS, basically any interaction with V8 handles is extremely slow. So we try to limit this as much as possible by encoding everything into the buffers.

Proper review soon - I've only given it a glance so far. 5% isn't much...

Comment (afinch7) [reply core/libdeno/binding.cc]: Ok I might encode this to control in Deno.core.dispatch, and decode in pre_dispatch.

Comment (afinch7) [reply core/libdeno/binding.cc]: I gave it a try and that seems to make things even slower not sure why though.

File: core/isolate.rs [hunk 5]
```diff
@@ -182,17 +191,34 @@ impl Isolate {
     let isolate = unsafe { Isolate::from_raw_ptr(user_data) };
     let control_shared = isolate.shared.shift();
 
-    let op = if control_argv0.len() > 0 {
+    let (op, promise_id, is_sync) = if control_argv0.len() > 0 {
       // The user called Deno.core.send(control)
       if let Some(ref f) = isolate.config.dispatch {
-        f(control_argv0.as_ref(), PinnedBuf::new(zero_copy_buf))
+        let promise_id =
+          i32::from_ne_bytes(control_argv0[..4].try_into().unwrap());
```
Comment (ry) [inline core/isolate.rs]: I don't like that the control buffer is no longer opaque to core. This constraints how op implementors do serialization.

There are a couple of distinct interface problems that are trying to be achieved here:
1. Allow `Op` to be generic on the Error.
1. Pass `is_sync` into the dispatch function.

I think the first item is achieved. Consider submitting it as a separate PR, I think we can land that. (Although I have a question about the performance of converting `Op::Async<E>` to `Op::Async<()>`, it seems the `Box<Future>` must be unwrapped to handle the error and serialize further.)

Regarding `is_sync`: 

I much prefer how you had it before, where `is_sync` (or `op_id`) actually becomes a parameter of `Deno.core.dispatch`. Having core inspect the payload of buffers is breaking the abstraction layer.

There is an incongruence between the op request and response: the JS side could as for a sync op and receive an async op. Or vise-versa. You've solved this by adding asserts, but if we more deeply refactor this interface we maybe could remove the asserts and solve the problem of having to add a parameter to `Deno.core.dispatch()`.

What if the requester does not specify if it wants sync or async? The requester somehow must encode what they desire into the control buffer - and the op creator returns whatever. When the JS side reads the response, it determines how to deal with it. This could be achieved with the simple convention: if `Deno.core.dispatch()` returns `null`, then the op is async. If not, the JS side can assume it is sync. The JS side can then do the asserts to check that the response matches what they requested, and core is not effected.

Comment (afinch7) [reply core/isolate.rs]: > There is an incongruence between the op request and response: the JS side could request a sync op and receive an async op. Or vise-versa. You've solved this by adding asserts on the Rust side, but if we more deeply refactor this interface we maybe could remove the asserts and solve the problem of having to add a parameter to Deno.core.dispatch().

Maybe if the rust side responds with a promise id then we know to expect a promise, and if it's not a number(a buffer) then we know the op is complete. Then it really doesn't matter how we respond unless the js side is expecting something specific, and in that case just encode it into the control.

Comment (ry) [reply core/isolate.rs]: Is it necessary to know the promise id?

File: cli/ops.rs [hunk 6]
```diff
@@ -992,7 +986,7 @@ fn op_open(
       ))
     });
   if is_sync {
-    let buf = op.wait()?;
+    let buf = tokio_util::block_on(op)?;
```
Comment (ry) [inline cli/ops.rs]: tokio_util::block_on should be avoided (very expensive). Use op.wait().

File: core/libdeno/deno.h [hunk 7]
```diff
@@ -84,7 +84,7 @@ void deno_execute(Deno* d, void* user_data, const char* js_filename,
 // releasing its memory.)
 //
 // If a JS exception was encountered, deno_last_exception() will be non-NULL.
-void deno_respond(Deno* d, void* user_data, deno_buf buf);
+void deno_respond(Deno* d, void* user_data, deno_buf buf, int* promise_id);
```
Comment (ry) [inline core/libdeno/deno.h]: Where does promise_id come from? Update docs in deno.h

Comment (afinch7) [reply core/libdeno/deno.h]: It comes from pre_dispatch. When the dispatch returns a Op::Async we create a promise id in rust and immediately return it via response/deno_respond. This means that a send/dispatch call will return a `Uint8Array | null | number` value. With numbers always being promise ids. I'll make sure to update docs to reflect this change.

**Comment by ry** [comment]
I wonder if you could add a method to isolate to get a `Vec<PromiseId>` of the current pending promises? Is that easy?

**Comment by afinch7** [comment]
> I wonder if you could add a method to isolate to get a `Vec<PromiseId>` of the current pending promises? Is that easy?

Easy? I don't see any reason it would be too hard. Not sure what the point would be though.

**Comment by ry** [comment]
I want to add a "Deno.top()" op at some point to see all the currently running ops.

Not a feature for this PR - just curious

**Comment by afinch7** [comment]
promiseMap.keys()?

**Comment by ry** [comment]
> promiseMap.keys()?

yes, that's good actually.

Regarding this patch - I want to land it - it's mostly good. However I am not convinced we need to have `is_sync` in the dispatch function. I think that should be handled by the op itself.

File: core/isolate.rs [hunk 8]
```diff
@@ -1073,7 +1125,7 @@ pub mod tests {
         let control = new Uint8Array(100 * 1024 * 1024);
         let response = Deno.core.dispatch(control);
         // Async messages always have null response.
```
Comment (ry) [inline core/isolate.rs]: Remove/adjust this comment

File: core/libdeno/deno.h [hunk 9]
```diff
@@ -81,8 +81,9 @@ void deno_execute(Deno* d, void* user_data, const char* js_filename,
 // deno_respond sends up to one message back for every deno_recv_cb made.
 //
 // If this is called during deno_recv_cb, the issuing libdeno.send() in
-// javascript will synchronously return the specified buf as an ArrayBuffer (or
-// null if buf is empty).
+// javascript will synchronously return the specified promise_id or buf
+// if promise_id is null as an ArrayBuffer (or null if buf and promise_id
+// are both null/empty).
```
Comment (ry) [inline core/libdeno/deno.h]: This text is confusing... Can you rephrase it in terms of async/sync ops?

File: core/shared_queue.js [hunk 10]
```diff
@@ -151,14 +151,27 @@ SharedQueue Binary Layout
...
-        asyncHandler(buf);
+        handleAsyncMsgFromRustInner(buf);
       }
     }
   }
 
+  function handleAsyncMsgFromRustInner(buf) {
+    // DataView to extract cmdId value.
+    const dataView = new DataView(buf.buffer, buf.byteOffset, 4);
+    const promiseId = dataView.getInt32(0);
```
Comment (ry) [inline core/shared_queue.js]: Ah... This is still bothering me because the buf is not opaque to core. This still defines some byte layout in the buffers.

The abstraction I want to make in //core is that these buffers are completely defined and controlled by the embedder and that core only passes them around.

Fixing this would require passing the promiseId as a parameter to the callback...

Comment (afinch7) [reply core/shared_queue.js]: ...and the best way to achieve this when using the shared buffer to pass results is to put this data into the record. I can't really think of any better way to do this off of the top of my head, but maybe it makes more sense to decode this information from the record in `shift()`.

Comment (ry) [reply core/shared_queue.js]: Oh! So this is only on the shared buffer? I think I might not have understood this correctly.

As far as the interfaces in core/isolate.rs, are the buffers passed to ops opaque?

Comment (afinch7) [reply core/shared_queue.js]: Yes promise_id is just appended to the beginning of a records when they are added to the shared buffer via `push()`. If you look down another line you can see we shrink the window used for the returned buffer, so that the promise_id in the record  is not directly exposed to non-core js code.

File: js/dispatch.ts [hunk 11]
```diff
@@ -56,14 +50,26 @@ function sendInternal(
...
+  isSync: true
+): Uint8Array | null;
+function sendInternal(
+  builder: flatbuffers.Builder,
+  innerType: msg.Any,
+  inner: flatbuffers.Offset,
+  zeroCopy: undefined | ArrayBufferView,
+  isSync: false
+): Promise<msg.Base>;
+function sendInternal(
```
Comment (ry) [inline js/dispatch.ts]: I've never seen this syntax... What does this mean?

File: core/shared_queue.rs [hunk 12]
```diff
@@ -152,17 +153,19 @@ impl SharedQueue {
     Some(&self.bytes[off..end])
   }
 
-  pub fn push(&mut self, record: &[u8]) -> bool {
+  pub fn push(&mut self, promise_id: c_int, record: &[u8]) -> bool {
```
Comment (ry) [inline core/shared_queue.rs]: What happens if there is no room in the shared_queue ?

File: core/shared_queue.js (hunk 10 above)
Comment (ry) [reply core/shared_queue.js]: 👍

File: core/libdeno/api.cc [hunk 13]
```diff
@@ -153,11 +153,14 @@ void deno_pinned_buf_delete(deno_pinned_buf* buf) {
...
   if (d->current_args_ != nullptr) {
     // Synchronous response.
-    if (buf.data_ptr != nullptr) {
+    if (promise_id != nullptr) {
+      auto number = v8::Number::New(d->isolate_, *promise_id);
+      d->current_args_->GetReturnValue().Set(number);
+    } else if (buf.data_ptr != nullptr) {
       auto ab = deno::ImportBuf(d, buf);
       d->current_args_->GetReturnValue().Set(ab);
     }
```
Comment (ry) [inline core/libdeno/api.cc]: Would it be appropriate to add the following? 
```c
} else {
  assert(false, "unreachable");
}
```

File: js/dispatch.ts (hunk 11 above)
Comment (afinch7) [reply js/dispatch.ts]: Function overloads

File: core/shared_queue.rs (hunk 12 above)
Comment (afinch7) [reply core/shared_queue.rs]: It gets passed here
https://github.com/afinch7/deno/blob/608d6c571af166ba52db0ad5b82c4d608ea7e2b4/core/isolate.rs#L618-L623

File: core/libdeno/api.cc (hunk 13 above)
Comment (afinch7) [reply core/libdeno/api.cc]: yes
